[Github-Actions-Workflows][Plugin-Release] Allow shipping a point-release for an older stable release

fullofcaffeine opened the pull request:
## What?

Allows the creation + release (as a`tag` in the WP SVN repo) of a point release for a given stable release, even when a newer (leaf/`latest`) stable release is present. The release should not be marked as `latest` when publishing and it will then run through a fork of the upload step that will ship it as a branch (under the `branches` directory) in the WP plugin repo (+ tag it under `tags`).

## Why?

Currently, if (a) bug(s) is(are) found in a given version of Gutenberg and a fix is worked on, we can then ship this fix(es) as a patch release for that version. However, sometimes during the window of time the bug was reported/fix implemented, **it could happen that the next stable release is shipped and that the next stable release isn't suitable to be released due to other newly-introduced issues. This effectively prevents a point release from being released for the previous stable release (which is not the `latest` anymore at this point). In the best case, the user must wait for the next version, which might take longer due to new issues that need to be worked out**. This can be a dangerous proposition as a new version might bring more issues and needs further testing, especially in big multisite WordPress instances like Wordpress.com, effectively creating a significant bottleneck for shipping the fix(es). Because of that, **we should be able to ship patch releases for older stable releases, even if a new stable release is already out, to have a faster turnaround when shipping bug-fixes. That's the goal of this changeset.**

## How?

- By allowing version bumping for old stable releases even if a new stable release is already out in `build-plugin-zip.yml` ([#](https://github.com/WordPress/gutenberg/pull/49082/files#diff-40b47d855e2cffe931290bba37c1b8716642627cbe05ac859728cb95c5da7536R59))
- By adding a new upload task to `upload-release-to-plugin-repo.yml` ([#](https://github.com/WordPress/gutenberg/pull/49082/files#diff-54a0a47c3496a3f17fce904a8bb01362b517363222210779f7dadb4e8fa2edd3R178)) that uploads that patch release as a branch (and tags it) to the WP plugins repo SVN. In this case, it leaves out `trunk` untouched in the SVN repo. **This step only runs if the `latest` stable (fetched using the GH API) differs from the published version**; otherwise, the `upload as trunk` runs instead. For this to work well, the GitHub draft release for this patch-release should NOT be marked as the `latest` release when it's published, or it will be considered the last release by the workflow and will replace `trunk`!*[0]
- It shouldn't affect the other existing scenarios.

_*[0] I don't see this as a huge problem, as long as we document this well. Also, this will often be done by people that understand the workflow and only in cases when an older patch release is needed when a new stable version is out, which shouldn't be too common (but still critical to ship when it happens!). Finally, if it happens that a patch-release for an older stable version ends in `trunk` because someone forgot to uncheck `Set as the latest release` when publishing, we can always revert the changes in the SVN repo manually._


## Testing Instructions

Testing this isn't very pleasant, but possible. You'll need `ngrok` (or equivalent tunnel tool) and a local web-dav-enabled svn server exposed via the tunnel. I've used this one: https://hub.docker.com/r/elleflorio/svn-server.

1. Fork the Gutenberg repo (not this repo). Make sure to recreate all the secrets from here. The ones needed are the SVN server username and password; the others are optional. Your new fork will not have any releases on the `releases` page, don't be surprised. 
2. Create a new release for the latest release tag, manually. I.e if the latest release for GB was 15.4.1, create a new release that points to this tag and is `latest`.
3. [Apply this diff](https://github.com/WordPress/gutenberg/pull/49082/commits/8cd123127c360a5035e2345960d913bbec74b280) to `trunk` and the the latest release branch (i.e `release/15.4`), and change the `PLUGIN_REPO_URL` to your tunnel URL.
4. Ensure your svn server is running and exposed to the web!

### Scenarios (happy paths)

#### Stable point release for the `latest` release
1. Go to `Actions->Build Gutenberg Plugin Zip`, leave `trunk` as the branch, and in the `rc or stable?` input box, type `stable`.
2. Once the action finishes, go to the `/releases` page, you should see a draft for the new patch release; click the pencil icon to edit it and then `Publish`. You should leave `Set as the latest release` checked here.
3. Go to actions again and open the `Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo ` (you will need to click `Show more workflows` in the sidebar to see it). 
4. Once it finishes, you should see the release published under `/releases` as `latest` and also in your SVN repo under `trunk` and under `tags/$VERSION` where `$VERSION` is the bumped version for this patch release.

#### Create a RC in order to create a new stable release

Let's create an RC to create a new stable release later:

1. Go to `Actions->Build Gutenberg Plugin Zip`, leave `trunk` as the branch, and in the `rc or stable?` input box, type `rc`.
2. Once the action finishes, go to the `/releases` page, you should see a draft for the new patch release, click the pencil icon to edit it and then `Publish`. You should leave `Set as prerelease` checked.

Now we have a new release branch and are ready to create a new stable release!

#### Create and publish a new stable release

This will set the ground for the last test, which will be to ship a point release for your previous stable release.

1. Go to `Actions->Build Gutenberg Plugin Zip`, leave `trunk` as the branch, and in the `rc or stable?` input box, type `stable`.
2. Once the action finishes, go to the `/releases` page; you should see a draft for the new patch release; click the pencil icon to edit it and then `Publish`. You should leave `Set as the latest release` checked here.
3. Go to actions again and open the `Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo ` (you must click `Show more workflows` in the sidebar to see it). 
4. Once it finishes, you should see the release published under `/releases` as `latest` and also in your SVN repo under `trunk` and under `tags/$VERSION` where `$VERSION` is the new stable release.

#### Create and publish a patch release for the previous stable release

This is the scenario that has been implemented in this changeset. Let's go ahead and test it out.

1. Go to `Actions->Build Gutenberg Plugin Zip`, select your previous stable release branch (i.e `release/15.4` if the new one is `release/15.5`) as the branch, and in the `rc or stable?` input box, type `stable`.
2. Once the action finishes, go to the `/releases` page; you should see a draft for the new patch release; click the pencil icon to edit it and then `Publish`. **IMPORTANT: You msut UNCHECK `Set as the latest release` checkbox, or it will be considered a regular `latest` release and will be pushed to the `trunk` of the WP SVN repo!**
3. Go to actions again and open the `Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo` (you will need to click `Show more workflows` in the sidebar to see it). 
4. Once it finishes, you should see the release published under `/releases`. It won't be set as `latest`. It will also your SVN repo under `branches/$VERSION` and `tags/$VERSION` where `$VERSION` is the new stable release.

## Illustrative Screenshots

| ![Screenshot from 2023-03-14 18-58-20](https://user-images.githubusercontent.com/81248/225180710-a1ba281c-04c8-4011-8a04-4122e72e08f3.png) |
|---|
| _The `Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo` workflow for `latest` stable (point or not) releases._ |

| ![Screenshot from 2023-03-14 19-39-52](https://user-images.githubusercontent.com/81248/225182567-4d463051-1b84-49c3-871d-613d917e62fe.png) |
|---|
| _The `Update Changelog and upload Gutenberg plugin to WordPress.org plugin repo` workflow for `non-latest` stable point-releases._ |

| ![Screenshot from 2023-03-14 18-57-13](https://user-images.githubusercontent.com/81248/225176137-5c1de1c8-68d6-4bb2-8c54-bb4fbb00acd6.png) |
|---|
| _The "non-`latest`" stable point-releases are created as a branch in the SVN repo. It's also copied over to `tags`, too._ |


## TODO
* [ ] Document the new release scenario https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md
* [ ] Figure out if there is a way to make the `Set as the latest release` checkbox unchecked by default for the previous-release (non-latest) patch releases? Please look at step 2) of the last test scenario for further context.

fullofcaffeine committed:
Allow point-releases for release branches that are not the latest anymore

And other minor improvements

fullofcaffeine committed:
Revert debug changes

Revert changes I made in order to test end-to-end from a test fork.

fullofcaffeine commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -95,11 +112,13 @@ jobs:
 
     upload:
-        name: Upload Gutenberg Plugin
+        name: Publish as trunk (and tag)
```
As a minor improvement, I changed the name to something that more specifically describes what the task does in relation to the new twin task that has been added below (one uploads to `trunk`, and tags (under `tags`), the other upload it as a branch (under `branches`) and tags.

fullofcaffeine commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -130,8 +145,8 @@ jobs:
                   STABLE_TAG_PLACEHOLDER: 'Stable tag: V\.V\.V'
-                  STABLE_TAG: 'Stable tag: ${{ steps.get_previous_stable_version.outputs.stable_version }}'
-              run: sed -i "s/${STABLE_TAG_PLACEHOLDER}/${STABLE_TAG}/g" ./trunk/readme.txt
+              run: |
```
Before, `STABLE_TAG` was based on the value of `steps.get_previous_stable_version.outputs.stable_version`, which for me looked a bit confusing, given the `get_previous_stable_version` step -- we actually want the current latest version, and not the previous, and we can get that from the `VERSION` env variable, defined as an `env` variable for the job and passed as part of the payload for the `release` event.

fullofcaffeine commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -159,3 +174,57 @@ jobs:
+              env:
+                  STABLE_TAG_PLACEHOLDER: 'Stable tag: V\.V\.V'
+              run: |
+                  sed -i "s/$STABLE_TAG_PLACEHOLDER/Stable tag: $VERSION/g" ./branches/$VERSION/readme.txt
```
Ditto for https://github.com/WordPress/gutenberg/pull/49082/files#r1136364711.

noahtallen commented:
> For this to work well, the GitHub draft release for this patch-release should NOT be marked as the latest release when it's published, or it will be considered the last release by the workflow and will replace trunk!*[0]

> we can always revert the changes in the SVN repo manually.

This is my only major concern. While I definitely follow your reasoning, I also see it being a very easy step to miss, even for someone experienced with the workflow. And yet it could have big repercussions if it actually overwrites the trunk version in WordPress.org, because that impacts anyone installing the plugin on their own sites, not just us

I think at the very least, we need to avoid pushing to the SVN repo in this scenario, even if someone accidentally sets the latest release in GitHub. I imagine changing the latest release in GitHub is easier than reverting in SVN?

fullofcaffeine commented:
> I think at the very least, we need to avoid pushing to the SVN repo in this scenario, even if someone accidentally sets the latest release in GitHub. I imagine changing the latest release in GitHub is easier than reverting in SVN?

Yeah, I'll see if I can come up with a condition that will avoid the publish event from triggering the upload action if the GB version being published is from an older branch. A simpler alternative might be to set it up so that the `Set as the latest release` is not checked by default for those releases, but I'm not sure if that's possible.

ockham commented:
> A simpler alternative might be to set it up so that the `Set as the latest release` is not checked by default for those releases, but I'm not sure if that's possible.

It seems like there's a `make_latest` field in GH's REST API to control this ([see](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release)).

---

_Edit:_ Oh, the docs say

> Drafts and prereleases cannot be set as latest.

So maybe we cannot control it for a draft, and there's a "latest" checkbox that's preselected when people go to manually publish it? 😕

ockham commented:
Looks like this needs a rebase to account for https://github.com/WordPress/gutenberg/pull/45932 🙂

ockham commented:
> Looks like this needs a rebase to account for #45932 🙂

Done. I made sure to preserve the commits, and updated the PR desc to point the instructions to the newly rebased first commit.

ockham commented on .github/workflows/build-plugin-zip.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
                 github.event.inputs.version == 'stable'
               ) || (
-                endsWith( github.ref, needs.compute-stable-branches.outputs.current_stable_branch ) &&
+                github.ref != 'refs/heads/trunk' &&
```
Ideally, we should also prevent running the action from arbitrary branches 🤔 
Maybe we can limit to `release/` branches?

```suggestion
                startsWith( github.ref, 'refs/heads/release/' ) &&
```

fullofcaffeine replied:
Sounds good 👍🏻

ockham commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -159,3 +174,57 @@ jobs:
+              working-directory: ./branches
+              run: |
+                  svn copy "$PLUGIN_REPO_URL/branches/$VERSION" "$PLUGIN_REPO_URL/tags/$VERSION" -m "Tagging version $VERSION" \
+                   --no-auth-cache --non-interactive  --username "$SVN_USERNAME" --password "$SVN_PASSWORD"
```
I know I had mentioned in our Slack convo that the tags should be created from the corresponding branches. However, what I meant was ideally to have a `x.y` branch (e.g. `15.4`), which would be created at the time that `15.4.0` is released, and from which all point releases are created. (This would be to mimic the behavior we have here in the GB git repo, and also what WP Core is doing in its svn repo.) 

Among other things, this would somewhat reflect in svn that e.g. 15.4.2 is based on 15.4.1, which in turn is based on 15.4.0, etc. However, the idea is also partly based on my ignorance of how branching and tagging works in svn. 

Maybe we can get away without any branching at all? 🤔 Would it be possible to commit the release files straight to `/tags/` (i.e. without committing them to `/branches/`)? (After all, for the actual plugin code, git is our source of truth, so we don't need to go out of our way to mimic the branches and tags in the GB git repo.)

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/49082/commits/9dc4b3976e9873f13337069abf1f4543760df2ac

ockham commented:
> > I think at the very least, we need to avoid pushing to the SVN repo in this scenario, even if someone accidentally sets the latest release in GitHub. I imagine changing the latest release in GitHub is easier than reverting in SVN?
> 
> Yeah, I'll see if I can come up with a condition that will avoid the publish event from triggering the upload action if the GB version being published is from an older branch. A simpler alternative might be to set it up so that the `Set as the latest release` is not checked by default for those releases, but I'm not sure if that's possible.

Yeah, that's exactly what I had in mind -- I think we can de-risk this workflow quite a bit by making sure the pre-selected default is fine (unchecked in this case).

I've noticed that the [`create-release` action is deprecated and unmaintained](https://github.com/actions/create-release). We could consider replacing it by a few lines of GH API (octokit) code (written in JS!) using the [`github-script` action](https://github.com/actions/github-script). (I really hope that there's some field that will allow us to control the default value of that "latest" checkbox.)

ockham commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -159,3 +174,57 @@ jobs:
+              uses: actions/download-artifact@9782bd6a9848b53b110e712e20e42d89988822b7 # v3.0.1
+              with:
+                  name: changelog trunk
+                  path: branches/$VERSION
```
Not sure this is working: I have a `branches/$VERSION/changelog.txt` in my svn now. (I.e. with a folder that's literally named `$VERSION`.)

I think variable substitution works differently in GH workflows/YAML files -- normal shell substitution only works in `run:` ('cause that's normally executed in a shell 😅 ). Everywhere else, we need `${{ }}` for [expressions](https://docs.github.com/en/actions/learn-github-actions/expressions), and env [variables](https://docs.github.com/en/actions/learn-github-actions/variables) aren't directly available (but through `env.xyz`, if I'm not mistaken). The following should work though:

```suggestion
                  path: branches/${{ github.event.release.name }}
```

fullofcaffeine commented:
Oh, good catch! You are right, it is easy to confuse the current context
though :/

On Wed, Mar 22, 2023, 3:00 PM Bernie Reiter ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/upload-release-to-plugin-repo.yml
> <https://github.com/WordPress/gutenberg/pull/49082#discussion_r1145402941>
> :
>
> > +                  # do the magic here
> +                  curl -L -o gutenberg.zip $PLUGIN_URL
> +                  unzip gutenberg.zip -d branches/$VERSION
> +                  rm gutenberg.zip
> +
> +            - name: Replace the stable tag placeholder with the existing stable tag on the SVN repository
> +              env:
> +                  STABLE_TAG_PLACEHOLDER: 'Stable tag: V\.V\.V'
> +              run: |
> +                  sed -i "s/$STABLE_TAG_PLACEHOLDER/Stable tag: $VERSION/g" ./branches/$VERSION/readme.txt
> +
> +            - name: Download Changelog Artifact
> +              uses: ***@***.*** # v3.0.1
> +              with:
> +                  name: changelog trunk
> +                  path: branches/$VERSION
>
> Not sure this is working: I have a branches/$VERSION/changelog.txt in my
> svn now. (I.e. with a folder that's literally named $VERSION.)
>
> I think variable substitution works differently in GH workflows/YAML files
> -- normal shell substitution only works in run: ('cause that's normally
> executed in a shell 😅 ). Everywhere else, we need ${{ }} for expressions
> <https://docs.github.com/en/actions/learn-github-actions/expressions>,
> and env variables
> <https://docs.github.com/en/actions/learn-github-actions/variables>
> aren't directly available (but through env.xyz, if I'm not mistaken). The
> following should work though:
> ⬇️ Suggested change
>
> -                  path: branches/$VERSION
> +                  path: branches/${{ github.event.release.name }}
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/49082#pullrequestreview-1353468536>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAT2YD6L2ELNOH4KSRPBZDW5NR5FANCNFSM6AAAAAAV3CRFDE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fullofcaffeine commented:
Oh, thanks for the clarification. In that case, yes, I think we can just
create it under `tags`, as it might be enough for WoA to pick it up, too
(but I will double-check).

On Wed, Mar 22, 2023, 2:35 PM Bernie Reiter ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/upload-release-to-plugin-repo.yml
> <https://github.com/WordPress/gutenberg/pull/49082#discussion_r1145380816>
> :
>
> > +            - name: Commit the content changes
> +              working-directory: ./branches
> +              run: |
> +                  svn st | grep '^?' | awk '{print $2}' | xargs -r svn add
> +                  svn st | grep '^!' | awk '{print $2}' | xargs -r svn rm
> +                  svn commit -m "Committing version $VERSION" \
> +                   --no-auth-cache --non-interactive  --username "$SVN_USERNAME" --password "$SVN_PASSWORD"
> +
> +            - name: Create the SVN tag
> +              working-directory: ./branches
> +              run: |
> +                  svn copy "$PLUGIN_REPO_URL/branches/$VERSION" "$PLUGIN_REPO_URL/tags/$VERSION" -m "Tagging version $VERSION" \
> +                   --no-auth-cache --non-interactive  --username "$SVN_USERNAME" --password "$SVN_PASSWORD"
>
> I know I had mentioned in our Slack convo that the tags should be created
> from the corresponding branches. However, what I meant was ideally to have
> a x.y branch (e.g. 15.4), which would be created at the time that 15.4.0
> is released, and from which all point releases are created. (This would be
> to mimic the behavior we have here in the GB git repo, and also what WP
> Core is doing in its svn repo.)
>
> Among other things, this would somewhat reflect in svn that e.g. 15.4.2 is
> based on 15.4.1, which in turn is based on 15.4.0, etc. However, the idea
> is also partly based on my ignorance of how branching and tagging works in
> svn.
>
> Maybe we can get away without any branching at all? 🤔 Would it be
> possible to commit the release files straight to /tags/ (i.e. without
> committing them to /branches/)? (After all, for the actual plugin code,
> git is our source of truth, so we don't need to go out of our way to mimic
> the branches and tags in the GB git repo.)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/49082#pullrequestreview-1353436551>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAT2YH3EKDU34X2IMQXDL3W5NO7TANCNFSM6AAAAAAV3CRFDE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fullofcaffeine commented:
Sure thing :)

On Wed, Mar 22, 2023, 1:13 PM Bernie Reiter ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/build-plugin-zip.yml
> <https://github.com/WordPress/gutenberg/pull/49082#discussion_r1145291568>
> :
>
> > @@ -56,7 +56,7 @@ jobs:
>                  github.event.inputs.version == 'rc' ||
>                  github.event.inputs.version == 'stable'
>                ) || (
> -                endsWith( github.ref, needs.compute-stable-branches.outputs.current_stable_branch ) &&
> +                github.ref != 'refs/heads/trunk' &&
>
> Ideally, we should also prevent running the action from arbitrary branches
> 🤔
> Maybe we can limit to release/ branches?
> ⬇️ Suggested change
>
> -                github.ref != 'refs/heads/trunk' &&
> +                startsWith( github.ref, 'refs/heads/release/' ) &&
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/49082#pullrequestreview-1353309455>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAT2YHBSR72J3G5FGQMPADW5NFM5ANCNFSM6AAAAAAV3CRFDE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fullofcaffeine commented:
Thanks!

On Wed, Mar 22, 2023, 8:08 AM Bernie Reiter ***@***.***>
wrote:

> Looks like this needs a rebase to account for #45932
> <https://github.com/WordPress/gutenberg/pull/45932> 🙂
>
> Done. I made sure to preserve the commits, and updated the PR desc to
> point the instructions to the newly rebased first commit.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/49082#issuecomment-1479632958>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAT2YAII7Y2KHI6L3BN2OTW5MBXTANCNFSM6AAAAAAV3CRFDE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fullofcaffeine commented:
Awesome, thanks for looking into it and pointing it out!

On Mon, Mar 20, 2023, 2:51 PM Bernie Reiter ***@***.***>
wrote:

> A simpler alternative might be to set it up so that the Set as the latest
> release is not checked by default for those releases, but I'm not sure if
> that's possible.
>
> It seems like there's a make_latest field in GH's REST API to control
> this (see
> <https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release>
> ).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/49082#issuecomment-1476917491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAT2YCXBEOJ63ESCR2KEL3W5C7M3ANCNFSM6AAAAAAV3CRFDE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

fullofcaffeine commented:
Apologies for the oversight

On Wed, Mar 22, 2023, 3:06 PM Marcelo Serpa ***@***.***> wrote:

> Oh, good catch! You are right, it is easy to confuse the current context
> though :/
>
> On Wed, Mar 22, 2023, 3:00 PM Bernie Reiter ***@***.***>
> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In .github/workflows/upload-release-to-plugin-repo.yml
>> <https://github.com/WordPress/gutenberg/pull/49082#discussion_r1145402941>
>> :
>>
>> > +                  # do the magic here
>> +                  curl -L -o gutenberg.zip $PLUGIN_URL
>> +                  unzip gutenberg.zip -d branches/$VERSION
>> +                  rm gutenberg.zip
>> +
>> +            - name: Replace the stable tag placeholder with the existing stable tag on the SVN repository
>> +              env:
>> +                  STABLE_TAG_PLACEHOLDER: 'Stable tag: V\.V\.V'
>> +              run: |
>> +                  sed -i "s/$STABLE_TAG_PLACEHOLDER/Stable tag: $VERSION/g" ./branches/$VERSION/readme.txt
>> +
>> +            - name: Download Changelog Artifact
>> +              uses: ***@***.*** # v3.0.1
>> +              with:
>> +                  name: changelog trunk
>> +                  path: branches/$VERSION
>>
>> Not sure this is working: I have a branches/$VERSION/changelog.txt in my
>> svn now. (I.e. with a folder that's literally named $VERSION.)
>>
>> I think variable substitution works differently in GH workflows/YAML
>> files -- normal shell substitution only works in run: ('cause that's
>> normally executed in a shell 😅 ). Everywhere else, we need ${{ }} for
>> expressions
>> <https://docs.github.com/en/actions/learn-github-actions/expressions>,
>> and env variables
>> <https://docs.github.com/en/actions/learn-github-actions/variables>
>> aren't directly available (but through env.xyz, if I'm not mistaken).
>> The following should work though:
>> ⬇️ Suggested change
>>
>> -                  path: branches/$VERSION
>> +                  path: branches/${{ github.event.release.name }}
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/WordPress/gutenberg/pull/49082#pullrequestreview-1353468536>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAT2YD6L2ELNOH4KSRPBZDW5NR5FANCNFSM6AAAAAAV3CRFDE>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

fullofcaffeine committed:
Limit runs to `release/` branches

Co-authored-by: Bernie Reiter <ockham@raz.or.at>

fullofcaffeine committed:
Fix changelog directory interpolation

Co-authored-by: Bernie Reiter <ockham@raz.or.at>

fullofcaffeine committed:
Re-add debug changes for further testing

fullofcaffeine committed:
Create only the tag when releasing an older point release, do not create a branch anymore

fullofcaffeine commented:
> > A simpler alternative might be to set it up so that the `Set as the latest release` is not checked by default for those releases, but I'm not sure if that's possible.
> 
> It seems like there's a `make_latest` field in GH's REST API to control this ([see](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#create-a-release)).
> 
> _Edit:_ Oh, the docs say
> 
> > Drafts and prereleases cannot be set as latest.
> 
> So maybe we cannot control it for a draft, and there's a "latest" checkbox that's preselected when people go to manually publish it? confused

Yeah, I just tested this, and unfortunately, it looks like that's the case.

fullofcaffeine commented:
> So maybe we cannot control it for a draft, and there's a "latest" checkbox that's preselected when people go to manually publish it? confused
> Yeah, I just tested this, and unfortunately, it looks like that's the case.

If we can't set that checkbox to unchecked by default and someone creates the release as `latest`, I don't think any condition we add to the upload zip workflow would work, as the de-facto last release would the newly promoted/created release (which in this case would be wrong, but there would be no way to know it using the current logic, which is based on checking the latest release).

I guess we could do some version arithmetic condition in the `upload zip` workflow to skip uploading if the latest version is < than the version on `trunk` (?), but it would still leave the wrong version set as `latest`. Ultimately, we could possibly have some additional logic in the upload workflow to do the arithmetic and then set the correct version as the latest if the checkbox was left checked, but things start to be overly complex.

I'll try reaching out to github via their support or forums to see if there's a way around that, but if not, I'm not sure what to do (other than clearly documenting it, as I mentioned in the PR description). Let me know if you have any other thoughts! cc @ockham @noahtallen

noahtallen commented:
> I guess we could do some version arithmetic condition in the upload zip workflow to skip uploading if the latest version is < than the version on trunk (?), but it would still leave the wrong version set as latest

Just re-iterating what we discussed in the call:

- it's super risky for this behavior to possibly update the Gutenberg plugin as well
- we could get the current plugin version from WordPress.org -- I found this: `curl -s 'https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&request\[slug\]=gutenberg' | jq -r '.version'`

I think using the latest version from WordPress.org would be a lot better for the plugin zip upload workflow, assuming the workflow actually wants to get the latest fully published version rather than the latest GH release

ockham commented:
> I think using the latest version from WordPress.org would be a lot better for the plugin zip upload workflow, assuming the workflow actually wants to get the latest fully published version rather than the latest GH release

Agree 💯

fullofcaffeine commented:
> > I guess we could do some version arithmetic condition in the upload zip workflow to skip uploading if the latest version is < than the version on trunk (?), but it would still leave the wrong version set as latest
> 
> Just re-iterating what we discussed in the call:
> 
>     * it's super risky for this behavior to possibly update the Gutenberg plugin as well
> 
>     * we could get the current plugin version from WordPress.org -- I found this: `curl -s 'https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&request\[slug\]=gutenberg' | jq -r '.version'`
> 
> 
> I think using the latest version from WordPress.org would be a lot better for the plugin zip upload workflow, assuming the workflow actually wants to get the latest fully published version rather than the latest GH release

Thanks for summing it up! 🙇🏻

fullofcaffeine committed:
Implement guard against accidental trunk release of older point releases

This refactors the upload zip workflow and add some version arithmetic to make sure the right version is uploaded depending:
- what is the current version set as latest -- this signals that a trunk release could be imminent, but is not enough to define it
- what is the current version in the WP core repo (real latest released version) -- this is compared with the published version and if it's greater than this version + the version was set as latest in GH, then it is a trunk release
- If the conditions above are false, then it's a tag release.

See the source for more details.

fullofcaffeine commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -5,22 +5,50 @@ on:
+                        // Only update trunk *if* the ref from the published plugin matches the latest
+                        // version in GH (releases) *and* if this version is actually GREATER than the
+                        // version published in the WP plugins repo. If not, then it will update the tag.
+                        const shouldUpdateTrunk = githubRef === `refs/tags/${latestVersionInGithub}` && isLatestVersionInGithubMoreRecent;
```
This expression should make sure that even if the release (non-prerelease) that was published was marked as `latest`, it will NOT be released as trunk here because the version arithmetic will make sure that `trunk` is only updated if the version that was published is actually **greater** than the version in the WP core repo. This means that, in the hypothetical case someone was releasing an old point release and accidentally left the "Mark as latest' checked, this expression would evaluate to `false`, as the version published would be **lesser** than the version in the WP core plugins repo.

fullofcaffeine replied:
Disclaimer that, at the time I write this, this hasn't been fully tested yet, but I will start testing now and continue testing early next week.

fullofcaffeine committed:
Attempt to fix "fetch" already defined error when running the script in the GH actions env

fullofcaffeine committed:
Simplify by not using node-fetch, or it will require installing the npm

fullofcaffeine committed:
GH requires an user-agent

fullofcaffeine committed:
Improve var name and comments

fullofcaffeine committed:
Define func to set the output of the "compute-should-update-trunk" job and improve its name

fullofcaffeine committed:
The core object should already be available as part of the github-scripts package

fullofcaffeine commented:
@ockham @noahtallen I've implemented some logic changes to prevent old releases from being accidentally pushed to `trunk`. I've elaborated a bit on it [here](https://github.com/WordPress/gutenberg/pull/49082#discussion_r1167353896), let me know if you have any questions, but the gist is that it adds a safeguard that relies on the version in the WP core repo, if the version being published form the draft is >, then it will be pushed to `trunk,` if not, it will be pushed as a `tag.` This doesn't prevent GitHub from marking the release as `latest` if the "Set as the latest release" checkbox is checked _(this also means that the "Set as latest release" checkbox doesn't have any effects in the upload workflow anymore)_. Still, at least it will prevent the real damage: in case someone accidentally published an old release as `latest,` it will be set as the `latest` on the `/releases` page, **but it won't be published to `trunk` in the svn repo**. One can just re-set the right last release as `latest` to fix it ✅ 

Feel free to let me know if you have any thoughts about the implementation or if you can think of any edge cases or issues! I'll test a bit more tomorrow, but I already went through the two scenarios in the PR description - but now just leaving the "Set as latest" checked when publishing" - and they worked as expected.

fullofcaffeine commented:
> Document the new release scenario https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md

I'll add this in a separate PR and link back here when it's ready.

fullofcaffeine commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -5,10 +5,63 @@ on:
+                          fetchJson('https://api.wordpress.org/plugins/info/1.2/?action=plugin_information&request[slug]=gutenberg'),
+                        ]);
+
+                        const latestVersionInGithub = githubData.tag_name
```
I think that a minor improvement (for visibility) would be to fetch these two versions in different jobs that this job would depend on, to give more visibility during the workflow runs (if debug logging is enabled) 🤔

fullofcaffeine replied:
Actually, I found a logic mistake here. I'm still relying on the `latest` release on GH, and I should just rely on the published ref to compare against the version on core. Fixing it now, will submit a change that will also simplify the job not to use JS and will split it into a few more jobs to make debugging easier.

fullofcaffeine committed:
Rely only in the github ref from the published release and not the latest tag, restructure the workflow to make it easier to debug and simplify it by using shell commands/script instead of JS

fullofcaffeine committed:
Simplify further by moving the fetch core repo job to a step inside compute_should_update_trunk

fullofcaffeine committed:
Fix/improve step/job ids

fullofcaffeine committed:
Fix typo

fullofcaffeine committed:
Formatting fixes

Auto-format with the RedHat YAML extension in vscode.

fullofcaffeine committed:
Add missing outputs and a couple of other fixes

fullofcaffeine committed:
A couple more fixes and a simplification of the compute_latest_version step

fullofcaffeine commented:
This should be ready for another review pass now 😅, feel free to have a look when you have time :) 

That said, I'll test the main scenarios again tomorrow, just in case. Testing is currently painful, as changes to the code need to be replicated in all branches for the release versions and trunk, but mainly because the plugin-building workflow takes a lot time to finish running. To avoid this, I'll try an approach of deleting the GB-building step and generating a zipped mock file instead. I'll document this workaround tomorrow, as it could be useful in the future. Also, to properly test the scenario where an old release is published with the "Set as latest" checked, the new logic relies on the version in the WP core repo. The version must be < the version in the WP core repo, or you should mock the value returned by the `Fetch latest version in the WP core repo` in the `compute-should-update-trunk` job. 

I'll create a PR to document the new flow added to https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md. 

Meanwhile, if you have the time, you're free to do a review pass and let me know if you find something odd :). If you have any questions, let me know!

fullofcaffeine commented on .github/workflows/build-plugin-zip.yml:
```diff
@@ -71,7 +71,7 @@ jobs:
             - name: Checkout code
               uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
               with:
-                  token: ${{ secrets.GUTENBERG_TOKEN }}
```
I will revert this change before merging (from `GITHUB_TOKEN` -> `GUTENBERG_TOKEN`).

fullofcaffeine commented:
> > Document the new release scenario https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md
> 
> I'll add this in a separate PR and link back here when it's rea

Docs have been updated in the following PR: https://github.com/WordPress/gutenberg/pull/49968/.

fullofcaffeine committed:
Mock the build of the gutenberg.zip for debugging purposes, this commit should be reverted before merging!

fullofcaffeine commented:
> To avoid this, I'll try an approach of deleting the GB-building step and generating a zipped mock file instead. I'll document this workaround tomorrow, as it could be useful in the future.

Done in https://github.com/WordPress/gutenberg/pull/49082/commits/014e2c6168c6945029ce7b1ac2ffbf3525004d4a. It adds the necessary directory structure and files (+content) that the upload workflow needs in some of its steps, and skips other unnecessary steps. It makes testing much more pleasant because of a faster turnaround time. Needless to say (but still saying anyway :P), this should be reverted before merging!

fullofcaffeine commented:
Updated the happy path flow diagrams for the upload workflow. It's worth noting that the SVN repo is a test one I have locally ;)

**Current version in the WP repo  at the time I tested this: `v15.6.1`.**

**Uploading a (mocked) `v15.7.10` - the state of "Set as latest release"*[0] doesn't matter anymore, but in this case, I purposely left it unchecked:**
![upload-15 7 10](https://user-images.githubusercontent.com/81248/233522351-1f819798-a032-4e57-bdb7-d4cafa321f7b.png)

**Proof the (mocked) `v15.7.10` has been uploaded to SVN (since `readme.txt` already existed, that's why it still has all the content, differently from the tag version):**

![Screenshot from 2023-04-20 19-56-27](https://user-images.githubusercontent.com/81248/233523292-a61e12ad-76cf-4c64-b13f-375e30922c32.png)


**Uploading a (mocked) `v14.7.5` -  the state of "Set as latest release"*[0] doesn't matter anymore, but in this case, I purposely left it checked:**

![upload-14 7 5](https://user-images.githubusercontent.com/81248/233522443-6c814050-0bdc-44f7-a9ad-4fcc91f93a51.png)

**Proof the (mocked) `v14.7.5` has been uploaded to SVN:**

![Screenshot from 2023-04-20 19-57-18](https://user-images.githubusercontent.com/81248/233523351-2f8f44e5-73b2-4790-a627-356a0486af4d.png)

![Screenshot from 2023-04-20 19-57-06](https://user-images.githubusercontent.com/81248/233523353-6981d7de-e255-4006-bec5-00a9c1ff184b.png)

---
_*[0] If the checkbox state is inconsistent with the production release version, it won't have serious consequences. Nonetheless, the person handling the release must fix the "latest" version on the GitHub releases page upon detecting the mistake. To my knowledge, a temporary incorrect "latest" version on the GitHub releases page may cause some confusion, but should not lead to major issues._

fullofcaffeine committed:
Fix a couple of step names

fullofcaffeine commented:
Pre-merge TODO:
- [ ] Review pass/approval cc @ockham (whenever you have a chance! 🙇🏻 )
- [x] Revert changes in https://github.com/WordPress/gutenberg/pull/49082/commits/014e2c6168c6945029ce7b1ac2ffbf3525004d4a and https://github.com/WordPress/gutenberg/pull/49082/commits/489aa7010679b4832da8766f9349396e2907aac4 and https://github.com/WordPress/gutenberg/pull/49082/commits/fe8b33c6ad301a8ea1260f869268f0b0531680c4.
- [x] Fix conflict with `trunk`
- [ ] Merge

noahtallen commented on .github/workflows/build-plugin-zip.yml:
```diff
@@ -56,7 +56,7 @@ jobs:
                 github.event.inputs.version == 'stable'
               ) || (
-                endsWith( github.ref, needs.compute-stable-branches.outputs.current_stable_branch ) &&
+                startsWith( github.ref, 'refs/heads/release/' ) &&
```
any reason for this change?

fullofcaffeine replied:
Yes; the whole condition looked like this before this change:

```
github.event_name == 'workflow_dispatch' && (
  (
    github.ref == 'refs/heads/trunk' ||
    endsWith(github.ref, needs.compute-stable-branches.outputs.next_stable_branch)
  ) && (
    github.event.inputs.version == 'rc' ||
    github.event.inputs.version == 'stable'
  ) || (
    endsWith(github.ref, needs.compute-stable-branches.outputs.current_stable_branch) &&
    github.event.inputs.version == 'stable'
  )
)
```

The first part until the last `||` binary op, deals with current or next releases:
* If the branch is `trunk` and the input is `rc/stable` it will evaluate to `true`. That will proceed to other steps that will check the input again to decide what to do next (i.e create a new branch if RC, or switch to it if stable. 

_I'm not entirely sure about the `endsWith( github.ref, needs.compute-stable-branches.outputs.next_stable_branch )` expression, but I think it's there to allow the user to trigger the workflow either from `trunk` or from the latest release branch, i.e make them equivalent (so, let's say that the last release is 15.6), triggering the workflow from `trunk` or `15.6` would yield the same result, because it's the tip of releases._

As you might now, the main use cases for releases are, which are documented [here](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/release.md), are:
* `trunk` + `stable` will result in a new non-RC stable release if the version logged in the branch's `package.json` (current ref) is an RC.
* `trunk` + `stable` will result in a patch release if the version in the branch's `package.json` is stable.
* `trunk` + `rc` will result in an RC for a new minor/major ([it depends](https://github.com/fullofcaffeine/gutenberg/blob/trunk/.github/workflows/build-plugin-zip.yml#L89-L92)) and will also create a new release branch.

This covers the conditions until the last OR operator. Now onto this expression:

```
(
    endsWith( github.ref, needs.compute-stable-branches.outputs.current_stable_branch ) &&
    github.event.inputs.version == 'stable'
)
```

This deals with the cases when the releaser selected a branch other than `trunk` _or_  the last release branch (the tip of the list of versions). This was there to allow point releases for the current stable release. As long as the next release was an RC, this expression would return true as the `current_stable_branch` would be the same as the current ref, [see this snippet](https://github.com/fullofcaffeine/gutenberg/blob/trunk/.github/workflows/build-plugin-zip.yml#L30-L44) - it gets the latest release from the GH API. However, if there was a more recent stable release, and if you were releasing for a previous version branch, then `current_stable_branch` would point to the more recent stable (`latest` in GH), and this would evaluate to `false`, as it would not match the branch ref where the workflow is running from. The `bump-version` step would be skipped, and with it, the whole workflow, as all the subsequent tasks depend on it.

That leads us to the change in this PR:

```
(
    startsWith( github.ref, 'refs/heads/release/' ) &&
    github.event.inputs.version == 'stable'
)
```

This was done to allow this workflow to be triggered for any previous release branches. This means the user can now run this for any past branch and not only for the `current_stable_branch`. Of course, this wasn't the only change needed for point releases for previous branches to work, but it was the first change required to allow this first workflow to be triggered for any past branches. Most of the logic changes are in the other workflow (`upload-release-to-plugin-repo.yml`).

fullofcaffeine replied:
And yes, it's a bit convoluted. I had to re-read this many times before writing the explanation above. Let me know if you have any other questions, as I might have missed something, too!

ockham replied:
Bit late to this, but your explanation is spot-on, @fullofcaffeine. Makes me wonder if we should add some comments to the expression (or those would just make it harder to parse 😅 )

> _I'm not entirely sure about the `endsWith( github.ref, needs.compute-stable-branches.outputs.next_stable_branch )` expression, but I think it's there to allow the user to trigger the workflow either from `trunk` or from the latest release branch, i.e make them equivalent (so, let's say that the last release is 15.6), triggering the workflow from `trunk` or `15.6` would yield the same result, because it's the tip of releases._

That's exactly right! My guiding principle for these workflows was to be as forgiving as possible to the user, as long as their intention could be ascertained. In this case, the intention is clear enough: they want to trigger a stable release for the current branch, and they shouldn't have to worry whether to select the `release/` branch (as they're familiar with from publishing point releases for previous release branches), or `trunk` -- it should "just work" either way.

noahtallen commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -5,10 +5,58 @@ on:
+              env:
+                  GITHUB_REF: ${{ github.ref }}
+              run: |
+                  latestPublishedVersion=$(echo "$GITHUB_REF" | sed -E 's/refs\/tags\/(v?)([0-9.]+)/\2/')
```
Where does the ref come from, and how does that connect with the latest version??

fullofcaffeine replied:
This workflow is triggered by publishing a draft release. This is the ref from the draft release, hence the name `lastPublishedVersion` for the variable -- it is the version (ref) of the last published release (previously-draft) created by the build zip workflow.

noahtallen commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -95,25 +144,23 @@ jobs:
-        if: ${{ !github.event.release.prerelease && github.event.release.assets[0] }}
+        needs: [compute-should-update-trunk, update-changelog]
+        if: |
+            needs.compute-should-update-trunk.outputs.should_update_trunk == 'true' &&
```
Nice! 👍

noahtallen commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -159,3 +206,51 @@ jobs:
                   svn commit -m "Releasing version $VERSION" \
                    --no-auth-cache --non-interactive  --username "$SVN_USERNAME" --password "$SVN_PASSWORD"
+
+    upload-tag:
```
How does this workflow get triggered?

fullofcaffeine replied:
Are you referring to the entire workflow or this specific job?

The workflow is triggered when a draft release is published (see [#](https://github.com/WordPress/gutenberg/blob/dda3c460c4a5c022ce18883f86fa94f115a9849b/.github/workflows/upload-release-to-plugin-repo.yml#L5)). This specific job only runs if the expression [here](https://github.com/WordPress/gutenberg/pull/49082/files/dda3c460c4a5c022ce18883f86fa94f115a9849b#diff-54a0a47c3496a3f17fce904a8bb01362b517363222210779f7dadb4e8fa2edd3R215-R217) evaluates to `true`.

noahtallen reviewed (commented):
nice, looking pretty good! I'll let @ockham or someone else give the final review. Just left a couple questions about how it works

fullofcaffeine committed:
Re-add prelease attribute, was accidentally commented out

fullofcaffeine committed:
Debug commit, please revert: mock the version number in the version repo

fullofcaffeine committed:
Fix changelog fetching when uploading the SVN tag

fullofcaffeine committed:
Mock SVN version to 16.1.0

fullofcaffeine committed:
Make the svn commit command verbose in order to debug network timeout issues while testing

fullofcaffeine committed:
Revert "Make the svn commit command verbose in order to debug network timeout issues while testing"

This reverts commit da46abafed451ca861fa082787f8d18eed093914.

fullofcaffeine committed:
Optimize SVN tag addition in GitHub workflow

This commit significantly optimizes the process of adding a new tag to the SVN repository in our GitHub Workflow.

Previously, our process involved checking out the entire tags directory from the SVN repository, which led to increased execution time and failure points when the tags directory was filled with numerous versions.

We've resolved this issue by leveraging the svn import command, which allows us to directly add a new directory to the tags directory without checking out the existing directories first. This results in faster and more reliable workflow execution, particularly when dealing with repositories that have accumulated many tagged versions.

The changes made include:

    Removing the step to checkout Gutenberg tags from the WP.org plugin repository.
    Adding a new step that uses the svn import command to directly add a new version directory to the SVN repository. This is achieved without the need to checkout all existing directories under tags/.
    Removing the step to checkout the new Gutenberg tag as it's not needed anymore.

fullofcaffeine committed:
Enhance safety checks in workflow to skip upload jobs for RC builds

This commit enhances the conditional checks within the 'upload' and 'upload-tag' jobs in our GitHub Actions workflows.

Previously, the jobs were executed based solely on the 'prerelease' flag of the release. However, there could be situations where this flag is accidentally set to 'false' for a Release Candidate (RC).

To avoid such mishaps, the conditions have been extended to check for the '-rc' suffix in the version string as well. This ensures that even if the 'prerelease' flag is incorrectly marked as 'false', the 'upload' and 'upload-tag' jobs are still skipped for RCs, thus preventing any unintended release of RC builds.

fullofcaffeine committed:
Revert "Enhance safety checks in workflow to skip upload jobs for RC builds"

This reverts commit eae536a6eeca7e7b7373361f451bdbfc58d45d10.

fullofcaffeine committed:
Skip upload jobs if release is an RC by checking the ref for a rc suffix

This makes sure we don't upload RCs as trunk or tag even if the deployer accidentally unchecks the "Set as prerelease" flag.

fullofcaffeine committed:
Revert build plugin mocking logic

Effectively reverts changes in 014e2c6.

fullofcaffeine committed:
Revert initial dev/debug changes

Reverts changes from 489aa70. If further testing is needed, revert the changes here again.

fullofcaffeine committed:
Revert WP SVN repo version mock

Revert changes from fe8b33c.

ockham commented on .github/workflows/upload-release-to-plugin-repo.yml:
```diff
@@ -5,10 +5,66 @@ on:
+                  latestPublishedVersion=$(echo "$GITHUB_REF" | sed -E 's/refs\/tags\/(v?)([0-9.]+)/\2/')
+                  latestVersionInCoreRepo="${{ steps.compute_latest_version_in_core_repo.outputs.version }}"
+
+                  # Determines if the first version string is greater than the second version string.
+                  #
+                  # Params:
+                  #   $1 - The first version string to compare, which may have an optional leading "v".
+                  #   $2 - The second version string to compare, which may have an optional leading "v".
+                  #
+                  # Return values:
+                  #   0 - The first version string is greater than the second version string.
+                  #   1 - The first version string is less than or equal to the second version string.
+                  is_first_version_greater_than_second() {
+                    v1=${1#v}
+                    v2=${2#v}
+                    dpkg --compare-versions "$v1" gt "$v2"
+                    return $?
+                  }
```
Definitely not for this PR, but in a follow-up, can we maybe replace this with the [`semver` npm](https://www.npmjs.com/package/semver)? I'm already using it in a few places in the build-plugin workflow ([e.g.](https://github.com/WordPress/gutenberg/blob/1acd27e54045130bc9b3dcc0ed79a781db251056/.github/workflows/build-plugin-zip.yml#L82)) to compute version numbers, and it also supports comparison

```
semver.gt('1.2.3', '9.8.7') // false
semver.lt('1.2.3', '9.8.7') // true
```

(and cleanup: `semver.clean('  =v1.2.3   ') // '1.2.3'`).

(`dpkg` is a Debian/Ubuntu tool IIRC, and I've tried to avoid bash-isms and other dependencies as much as possible in order to make local testing of individual steps from workflows at least a bit easier across platforms -- including MacOS.)

fullofcaffeine replied:
That's a good point. I'll take note of that for a follow-up 👍🏻

ockham reviewed (approved):
Okay. Based on all the testing we've done, plus the safeguards you've added, LGTM 🤞😮‍💨🚢

ockham commented:
Let's just make sure to announce this properly, e.g. in `#core-editor`. Thinking of something like:

> @fullofcaffeine and I would like to land a [PR](https://github.com/WordPress/gutenberg/pull/49082) that adds a new feature to the Gutenberg release workflow (in GitHub Actions), which will enable us to publish a point release for a previous version series even if a new stable release has already been published. E.g. if the current Gutenberg version is 15.9.1, and we discover a serious bug in 15.8.1, Marcelo’s change will allow us to publish a 15.8.2 release with a bugfix — somewhat akin to Core. (In this example, 15.9.1 will continue to be marked as 'latest' release, so no WordPress installs running it will be at risk of downgrading.)
>
> The PR contains a number of safeguards and has undergone exhaustive testing, but due to the way that GHA workflows work, and due to the many parts involved, there’s still a risk that we missed some edge-case bug. For this reason, we’ll make sure that at least one of us will be around when the next Gutenberg RC and stable versions are published. 

Am I missing anything important? 🤔 

Maybe one of us should volunteer to run the first GB release after landing this 🤔 
Speaking of landing this: We should probably do that after a stable release is out (and before a new series is started by an RC). Let's maybe consult with @priethor on when would be a good time (also considering WP 6.3).

fullofcaffeine commented:
@ockham Sounds good! @priethor 👋🏻 When would be a good day to release this? At first glance, I'm thinking of scheduling to the next week (19th-23rd June, any of these days), but I'm not sure what's the 6.3 status/timeline. Thanks!

tellthemachines commented:
The next Gutenberg release (16.1) will be the last before 6.3 Beta 1, and as editor tech leads, either @ramonjd or myself will likely be running it. If this isn't an urgent change, it might be safer to aim for 16.2 so that there's no risk of disruption to 6.3 Beta.

ockham commented:
Sounds good, thank you @tellthemachines. So if I'm not mistaken:
- 16.0 is due tomorrow (June 14); which means that
- 16.1 is due on June 28, and
- 16.2 is due on July 12 (with an RC1 on July 5th).

@fullofcaffeine How about we merge this on July 5th and also take care of publishing 16.2 RC1 right afterwards?

fullofcaffeine commented:
> @fullofcaffeine How about we merge this on July 5th and also take care of publishing 16.2 RC1 right afterwards?

Sounds like a good plan! 👍🏻

fullofcaffeine committed:
Add back if conditions that were accidentally removed

Removed in https://github.com/WordPress/gutenberg/pull/49082/commits/014e2c6168c6945029ce7b1ac2ffbf3525004d4a.

I accidentally removed it instead of commenting out, so I forgot to reactivate it.

The problem was that due to the amount of changes later, I couldn't just revert this commit.
